Update client-go latency metrics bucket

aojea opened the pull request:
/kind cleanup

#### What this PR does / why we need it:


An example of current buckets used for latency shows the problem of capping it to .512 seconds, we loose visibility on the high values and I think we don't need such granularity on the low values

![image](https://user-images.githubusercontent.com/6450081/145389807-846ae77e-f832-441e-b0e3-b739ee7a7204.png)


```release-note
client-go metrics: change bucket distribution for rest_client_request_duration_seconds and rest_client_rate_limiter_duration_seconds from [0.001, 0.002, 0.004, 0.008, 0.016, 0.032, 0.064, 0.128, 0.256, 0.512] to [0.005, 0.025, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0, 15.0, 30.0, 60.0}]
```

aojea committed:
update buckets for client-go latency metrics

Current request latency metrics have the following buckets:

0.001, 0.002, 0.004, 0.008, 0.016, 0.032, 0.064, 0.128, 0.256, 0.512

That has two much granularity for http requests, and it gets capped to
aprox half seconds, loosing visibility on the requests that may be more
interested, the ones that take more than 1 second.

Using the same used for etcd request latency, with the same upper and
lower limits than the ones used in the apiserver, but only adding one
bucket more.

[]float64{0.005, 0.025, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 8.0, 15.0, 30.0, 60.0},

aojea commented:
/assign @deads2k @wojtek-t

aojea commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -926,12 +926,6 @@ func (r *Request) newHTTPRequest(ctx context.Context) (*http.Request, error) {
 // fn at most once. It will return an error if a problem occurred prior to connecting to the
 // server - the provided function is responsible for handling server errors.
 func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Response)) error {
-	//Metrics for total request latency
```
IIUIC this is only called from `Do()` and `DoRaw()`

aojea commented on staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:
```diff
@@ -40,6 +40,16 @@ var (
+			Name: "rest_client_request_size_bytes",
+			Help: "Request size in bytes. Broken down by verb and URL.",
+			// Use buckets ranging from 1000 bytes (1KB) to 10^9 bytes (1GB).
+			Buckets: k8smetrics.ExponentialBuckets(1000, 10.0, 7),
```
These are the same values used in the apiserver
```
staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:                  Buckets:        compbasemetrics.ExponentialBuckets(1000, 10.0, 7),
```

aojea commented on staging/src/k8s.io/client-go/go.mod:
```diff
@@ -24,14 +21,14 @@ require (
 	google.golang.org/protobuf v1.27.1
 	k8s.io/api v0.0.0
 	k8s.io/apimachinery v0.0.0
+	k8s.io/component-base v0.0.0
```
@liggitt is ok to import this, it is pulled by the example :shrug:

deads2k commented:
/cc @logicalhan

liggitt commented on staging/src/k8s.io/client-go/examples/metrics/main.go:
```diff
@@ -0,0 +1,82 @@
+	"time"
+
+	"k8s.io/component-base/metrics/legacyregistry"
+	_ "k8s.io/component-base/metrics/prometheus/clientgo" // load all the prometheus client-go plugin
```
I would not expect this dependency in this direction

liggitt replied:
not only would I not expect it, component-base depends on client-go, which makes this a circular dependency, which we don't want

dgrisonnet commented on staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:
```diff
@@ -143,6 +155,14 @@ func (l *latencyAdapter) Observe(ctx context.Context, verb string, u url.URL, la
+}
+
+func (s *sizeAdapter) Observe(ctx context.Context, verb string, u url.URL, size float64) {
+	s.m.WithContext(ctx).WithLabelValues(verb, u.String()).Observe(size)
```
Considering this recent cardinality explosion, we might prevent `u.Path` instead of `u.String()`: https://github.com/kubernetes/kubernetes/pull/106539

tkashem replied:
maybe `r.finalURLTemplate()`?

dgrisonnet replied:
If I am not mistaken, `r.finalURLTemplate()` might also contain the query string, so it won't solve the issue that the query string can take an unbounded number of values. In addition, we don't have access to the request here so we can't really call this method. What we should definitely do though is call `Observe` with `r.finalURLTemplate` and set `u.Path` and `u.Host` in the metric as mentioned by @aojea in https://github.com/kubernetes/kubernetes/pull/106539#discussion_r770328567

tkashem replied:
I would recommend keeping the labels and their values same for both metrics here - `RequestLatency` and `RequestSize`

aojea replied:
yeah, we have to be consistent there

dgrisonnet commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1035,18 +1029,24 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 	if err != nil {
 		return Result{err: err}
 	}
+	metrics.RequestLatency.Observe(ctx, r.verb, r.finalURLTemplate(), time.Since(start))
```
Do you know if `r.finalURLTemplate()` will return the whole URI? If that's the case, we should at least remove the query parameters since they can lead to cardinality issues. The best in my opinion would be to only have the path of the request since it should be good enough for debugging.

aojea replied:
yeah, I thought the same, a CI run in openshift returns 841 series by URL, seems excessive

https://xzrmvbis-promecieus.apps.cr.j7t7.p1.openshiftapps.com/graph?g0.expr=sum%20by%20(url)%20(rest_client_request_duration_seconds_bucket%7B%7D)&g0.tab=0&g0.stacked=0&g0.show_exemplars=0&g0.range_input=45m51s864ms&g0.end_input=2021-12-09%2016%3A45%3A57&g0.moment_input=2021-12-09%2016%3A45%3A57

However, the request host is important, we need to be able to filter by destination host ... from the client side maybe the path is less important?

dgrisonnet replied:
Indeed the host is important, I wrongly assumed that it would always be the apiserver. 

>  from the client side maybe the path is less important

I'd say that the path is as important since it allows granularity. When defining your SLOs, you might want to skip certain paths / define different SLOs depending on the path that the client is hitting.

So I guess what we want is a combination of the Scheme + Host + Path? Everything except the query params and the fragments really.

dgrisonnet replied:
@aojea following up on our offline discussion, I think we should have one label for the `host` and another for the `path`, so that we won't need to do any post computation to filter based on the path or the host.

aojea replied:
agree, but then I will wait for your PR to merge https://github.com/kubernetes/kubernetes/pull/106539

deads2k replied:
@aojea I think the kube-apiserver breaks apart the queries by resource and verb.  Does this method break it apart in the same way for comparison?

aojea replied:
we already have it
https://github.com/kubernetes/kubernetes/blob/15112c7286497cc42b5f0552b8dc5c935c6cac04/staging/src/k8s.io/client-go/rest/request.go#L791-L795

aojea replied:
I think that you refer to this

ttps://github.com/aojea/kubernetes/blob/1d9d530ee1b672acb9f2ba089123b5350d64dc3b/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L101-L110

 I don't have any inconvenient on trying to standardize and I think is a good idea to be consistent so we can reuse queries, ... , however, the only important field here is the `host:port` of the url, the `path` adds value from the client perspective if we can know that path is being proxied.

We are only interested in the time to get a response to understand if something between the client and the host is acting up. For any other analysis the apiserver audit.log or apiserver metrics has to be the authoritative answer.

dgrisonnet reviewed (commented):
The metrics look good from my side

aojea commented:
/hold

It seems there are some historical context needed

aojea commented on staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:
```diff
@@ -35,7 +35,7 @@ var (
 		&k8smetrics.HistogramOpts{
 			Name:    "rest_client_request_duration_seconds",
 			Help:    "Request latency in seconds. Broken down by verb and URL.",
-			Buckets: k8smetrics.ExponentialBuckets(0.001, 2, 10),
```
can this be this way to calculate the calculate the SLO using 512 ms?

dgrisonnet replied:
Do you mean with the previous version? From what I can tell the previous version's biggest bucket is 512ms so it is possible. But I would argue that `k8smetrics.DefBuckets` gives more flexibility since it goes up to 10s with one of its buckets at 500ms.

aojea replied:
Yeah , I was trying to find a good answer for.current buckets ... But I can't defend it, the default ones are much better

tkashem replied:
default buckets:
```
DefBuckets = []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10}
```
for non long running requests, the timeout enforced on the server is `60s` (if client does not specify any timeout). Do you think it is worth adding more buckets on the upper bound?

This is how it's defined in the apiserver:
https://github.com/kubernetes/kubernetes/blob/1e6f3b5cd68049a3501782af8ff3ddd647d0b408/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L106-L107

aojea replied:
is going to make any difference to know that latency is 20 or 25 or 45 to the SRE/admin/operator? 
I don't think so, same as having such granularity on lower bounds, if there is nothing actionable I don't see the value of having too much information

dgrisonnet replied:
I agree with Antonio here, plus I don't think that from a client perspective we care about long-running requests. We essentially went to get insight on everything in between the client and the server, so `GET` requests are enough.

tkashem replied:
i think it helps in debugging; during degradation, requests can take more than `10s`. Instead of seeing a flat line on 10s on the graph, it would be more revealing if we have a buckets for `{15, 20, 30, 45, 60}`, unless cardinality is a concern here.

aojea replied:
I'm going to use the same that the ones used in etcd then, that seems a common ground.
On the contrary to the apiserver, that  measure the latency to process a request and it maybe CPU only, this has always to measure the network roundtrip + the apiserver latency to process the request
```
[]float64{0.005, 0.025, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 15.0, 30.0, 60.0},
```

aojea replied:
this keeps the same number of buckets and also is able to track latencies within the same limits the apiserver is doing

fedebongio commented:
/assign @logicalhan 
/triage accepted

tkashem commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1035,18 +1029,24 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 		return Result{err: err}
 	}
+	metrics.RequestLatency.Observe(ctx, r.verb, r.finalURLTemplate(), time.Since(start))
+	metrics.RequestSize.Observe(ctx, r.verb, r.finalURLTemplate(), float64(len(result.body)))
```
can we put it in a `defer` block to maintain current behavior?

aojea replied:
current behavior is wrong, you can't measure latency if there is no response or if you didn't connect to the server, per example, if the request timed out because a network error, we were recording its latency

deads2k replied:
> current behavior is wrong, you can't measure latency if there is no response or if you didn't connect to the server, per example, if the request timed out because a network error, we were recording its latency

I can buy that.  Do I get a new metric measuring timeouts?

aojea replied:
we already have it
https://github.com/kubernetes/kubernetes/blob/15112c7286497cc42b5f0552b8dc5c935c6cac04/staging/src/k8s.io/client-go/rest/request.go#L791-L795

tkashem replied:
current behavior includes error (`err` is set) where the network peer has been reached and something went wrong while the request was being processed. i think we should include these in the metrics

https://github.com/kubernetes/kubernetes/blob/9918aa1e035a00bc7c0f16a05e1b222650b3eabc/staging/src/k8s.io/client-go/rest/request.go#L1024-L1026

aojea replied:
I think that err comes from

https://github.com/kubernetes/kubernetes/blob/9918aa1e035a00bc7c0f16a05e1b222650b3eabc/staging/src/k8s.io/client-go/rest/request.go#L980-L986

and the docs say https://pkg.go.dev/net/http#Client.Do
```
func (*Client) Do ¶
func (c *Client) Do(req *Request) (*Response, error)
Do sends an HTTP request and returns an HTTP response, following policy (such as redirects, cookies, auth) as configured on the client.

An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem). A non-2xx status code doesn't cause an error.
```

aojea replied:
> An error is returned if caused by client policy (such as CheckRedirect), or failure to speak HTTP (such as a network connectivity problem)

err is not something we want to use for measuring latency

aojea replied:
nevermind, I kept the same bahevaviour for the latency

aojea replied:
@tkashem you were right I was wrong, it seems the common practices is to track latency also for failing request , I removed the commit changing that

deads2k commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -1057,6 +1057,8 @@ func (r *Request) DoRaw(ctx context.Context) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	metrics.RequestLatency.Observe(ctx, r.verb, r.finalURLTemplate(), time.Since(start))
```
Do you plan to add a latency per unit size?

aojea replied:
I want to give the users/admins the opportunity to estimate bandwidth (bits per second), that will allow to easily identify devices that rate limit traffic, per example to 10 mpbs ;)

aojea committed:
client-go: remove no longer used finalURLTemplate

The restclient metrics were updated to track only the host field of the
url, the finalURLTemplate is not longer needed, its only goal was to
replace name and namespace in the path to avoid cardinality.

aojea commented:
/hold

please hold on reviews, I want to do a more exhaustive analysis and add some examples so you can observe the results.
I just found that we are taking the whole latency, including the retries ... and not measuring the number of retries per request , we should

aojea commented:
/test pull-kubernetes-verify

aojea commented:
/test pull-kubernetes-kubemark-e2e-gce-scale
/retest

marseel commented:
/test pull-kubernetes-e2e-gce-100-performance 
This run should contain prometheus snapshot from https://github.com/kubernetes/test-infra/pull/25348

aojea commented:
> /test pull-kubernetes-e2e-gce-100-performance This run should contain prometheus snapshot from [kubernetes/test-infra#25348](https://github.com/kubernetes/test-infra/pull/25348)

working like a charm @marseel :tada: 

![image](https://user-images.githubusercontent.com/6450081/154981805-9d8a9cf4-9701-4b19-90fc-d63363b9a37f.png)

marseel commented:
You may also find these dashboards for grafana quite useful: https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/pkg/prometheus/manifests/dashboards

aojea commented:
> You may also find these dashboards for grafana quite useful: https://github.com/kubernetes/perf-tests/tree/master/clusterloader2/pkg/prometheus/manifests/dashboards

those are awesome, I'm going to add the client metrics graphs  there then

aojea commented:
Yeah , absolutely , I was more asking about the process I should follow ... Thanks for your help

tkashem commented on staging/src/k8s.io/client-go/rest/with_retry.go:
```diff
@@ -123,6 +125,8 @@ func (r *withRetry) NextRetry(req *http.Request, resp *http.Response, err error,
 	}
 
 	r.attempts++
+	updateRetryURLMetrics(context.TODO(), req, resp, err)
```
I don't think that's the appropriate place for it, this function returns retry parameters, whether a retry should take place. For example, for a request that succeeds the first time, the retry metric would record one retry.

aojea replied:
yeah, that was my feeling, I was going to ask you about this because I wasn't sure
thanks for confirming

aojea replied:
removed commit from this PR

tkashem reviewed (commented):
I would recommend pulling out the retry metric in its own PR, since the retry metric is orthogonal to the other improvements you are making in this PR.

aojea commented:
this is how it looks like now

![image](https://user-images.githubusercontent.com/6450081/155182938-377b5f80-90e2-4207-baec-38f993ecd5b0.png)

aojea commented:
/assign @wojtek-t @tkashem 

This PR now only updates the bucket for latencies, so we are not limited to 512 msec.
It also removes the function finalURLTemplate that is not required since the metrics were updated to track only the host of the URL https://github.com/kubernetes/kubernetes/pull/106539

aojea commented:
/test pull-kubernetes-integration

job with timeout :/ https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/106911/pull-kubernetes-integration/1496390862123307008 
to the bucket of integration weirdness https://github.com/kubernetes/kubernetes/pull/108060#issuecomment-1048606973

dgrisonnet commented on staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:
```diff
@@ -35,7 +35,7 @@ var (
 			Name:    "rest_client_request_duration_seconds",
 			Help:    "Request latency in seconds. Broken down by verb, and host.",
-			Buckets: k8smetrics.ExponentialBuckets(0.001, 2, 10),
+			Buckets: []float64{0.005, 0.025, 0.1, 0.25, 0.5, 1.0, 2.0, 4.0, 15.0, 30.0, 60.0},
```
there is quite a big gap between 4s and 15s, so I think we should add one additional bucket to improve measurements. I would be fine with either 8s or 10s.

aojea replied:
donde

dgrisonnet reviewed (commented):
Could you perhaps mention the exact bucket change in the release note? It would make it easier to the users to make the transition.

For example, we add the following in 1.23:
```
Changed buckets in apiserver_request_duration_seconds metric from [0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,1.25, 1.5, 1.75, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 40, 50, 60] to [0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3, 4, 5, 6, 8, 10, 15, 20, 30, 45, 60]
```

wojtek-t commented on staging/src/k8s.io/client-go/rest/request.go:
```diff
@@ -601,7 +523,7 @@ func (r *Request) tryThrottleWithInfo(ctx context.Context, retryInfo string) err
 		globalThrottledLogger.Infof("%s", message)
 	}
-	metrics.RateLimiterLatency.Observe(ctx, r.verb, r.finalURLTemplate(), latency)
+	metrics.RateLimiterLatency.Observe(ctx, r.verb, *r.URL(), latency)
```
I have a feeling that this discussion already happened, but aren't we afraid of cardinality here?

aojea replied:
yeah, Damien replaced the metrics to only use host, path is no longer https://github.com/kubernetes/kubernetes/pull/106539 present

wojtek-t replied:
So shouldn't we use host here too?

aojea replied:
the interface takes an url.URL() 

https://github.com/kubernetes/kubernetes/blob/343125cc6c4a8b83b6fad8778d9fc1cd7ad1f947/staging/src/k8s.io/client-go/tools/metrics/metrics.go#L40-L43

and is an exported interface, is ok to change the signature?

wojtek-t replied:
aah ok - I misunderstood initially. So it's already fixed for this metric too already.

We can discuss changing the interface, but given it's a client one it's always better to avoid it (and even if not, it's not for this PR).

aojea replied:
:upside_down_face:

aojea commented:
> Could you perhaps mention the exact bucket change in the release note? It would make it easier to the users to make the transition.
> 
> For example, we add the following in 1.23:
> 
> ```
> Changed buckets in apiserver_request_duration_seconds metric from [0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0,1.25, 1.5, 1.75, 2.0, 2.5, 3.0, 3.5, 4.0, 4.5, 5, 6, 7, 8, 9, 10, 15, 20, 25, 30, 40, 50, 60] to [0.05, 0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.25, 1.5, 2, 3, 4, 5, 6, 8, 10, 15, 20, 30, 45, 60]
> ```

done

logicalhan reviewed (commented):
/lgtm

dgrisonnet commented:
/lgtm
